cpp/src_prims/stats/histogram.h

```diff
@@ -0,0 +1,488 @@
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+void histogramVecLen(HistType type, int* bins, IdxT nbins, const DataT* data,
+                     IdxT n, int tpb, cudaStream_t stream, BinnerOp op) {
+  CUDA_CHECK(cudaMemsetAsync(bins, 0, nbins * sizeof(int), stream));
+  switch (type) {
+    case HistTypeGmemWarp:
+      gmemWarpHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                                  stream);
+      break;
+    case HistTypeGmem:
+      gmemHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                              stream);
+      break;
+    case HistTypeSmemWarp:
+      smemWarpHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                                  stream);
+      break;
+    case HistTypeSmem:
+      smemHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                              stream);
+      break;
+    case HistTypeSmemBits16:
+      smemBitsHist<DataT, BinnerOp, IdxT, 16, VecLen>(bins, nbins, data, n, op,
+                                                      tpb, stream);
+      break;
+    case HistTypeSmemBits8:
+      smemBitsHist<DataT, BinnerOp, IdxT, 8, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemBits4:
+      smemBitsHist<DataT, BinnerOp, IdxT, 4, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemBits2:
+      smemBitsHist<DataT, BinnerOp, IdxT, 2, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemBits1:
+      smemBitsHist<DataT, BinnerOp, IdxT, 1, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemHash:
+      smemHashHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                                  stream);
+      break;
+    default:
+      ASSERT(false, "histogram: Invalid type passed '%d'!", type);
+  };
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT>
+void histogramImpl(HistType type, int* bins, IdxT nbins, const DataT* data,
+                   IdxT n, int tpb, cudaStream_t stream, BinnerOp op) {
+  size_t bytes = n * sizeof(DataT);
+  if (n <= 0) return;
+  if (16 / sizeof(DataT) && bytes % 16 == 0) {
+    histogramVecLen<DataT, BinnerOp, IdxT, 16 / sizeof(DataT)>(
+      type, bins, nbins, data, n, tpb, stream, op);
+  } else if (8 / sizeof(DataT) && bytes % 8 == 0) {
+    histogramVecLen<DataT, BinnerOp, IdxT, 8 / sizeof(DataT)>(
+      type, bins, nbins, data, n, tpb, stream, op);
+  } else if (4 / sizeof(DataT) && bytes % 4 == 0) {
+    histogramVecLen<DataT, BinnerOp, IdxT, 4 / sizeof(DataT)>(
```

Comment:
Will it work if `sizeof(DataT) == 3`, but `n % 4 == 0`, and thus `bytes % 4 == 0`?

Reply:
Most use-cases will be either 32b/64b int's or float/double. For all these cases, this check shouldn't have a problem. But you're right, better safe than to be sorry. See my previous response for the fix. Thanks.